### ISSUE TITLE
enhancement to GitHub Action testing for DB

### ISSUE THREAD

**Issue by huss** [OpenEnergyDashboard/OED#553]
In reference to the new setup for GitHub Actions, @campionfellin  said:
One nice benefit to a lot of these now is instead of having them in plaintext in the .travis.yml file, they can be put into the Github secrets section and then inserted like ${{ secrets.OED_DB_TEST_USER }}.

Then perhaps down the road a test postgres db can be set up permanently that all the tests connect to.

**Comment by huss** [comment]
These tests are run from .github/workflows/build.yml.

**Comment by corygreer18** [comment]
Niccm (Nick Mederos) and I worked on documentation and next steps for this issue that can be found in the below file.
[enhancingGitHubAction.md](https://github.com/OpenEnergyDashboard/DesignDocs/blob/main/githubAction/enhancementToGithubAction.md)

**Comment by knakatasf** [comment]
Is this issue still open? I'd like to work on this issue. I learned GitHub Actions through the coursework recently, so I might be able to help.

**Comment by huss** [comment]
@knakatasf Yes, it is available to work on and you are welcome to investigate this. I've assigned it to you. Please let me know if I can help.

**Comment by knakatasf** [comment]
Hi @huss, how are you doing?
I understand that we want to transfer the jobs currently handled by .travis.yml to .github/workflows/build.yml. If that's the case:
(1) I will update build.yml so it will build and run the web container alongside the database container. Currently, build.yml only runs the raw web program.
(2) build.yml will reference GitHub secrets for some credentials and pass those environment variables to docker-compose.yml.
Does that sound correct?

**Comment by huss** [comment]
@knakatasf Thank you for asking about this issue.

>I understand that we want to transfer the jobs currently handled by .travis.yml to .github/workflows/build.yml. 

My best understanding is that we no longer use the Travis system and only GitHub actions. The Travis files are historical and we used to use that system. I'm not expert here so if you think otherwise then please let me know.

>If that's the case: (1) I will update build.yml so it will build and run the web container alongside the database container. Currently, build.yml only runs the raw web program. (2) build.yml will reference GitHub secrets for some credentials and pass those environment variables to docker-compose.yml. Does that sound correct?

I agree with (2). We may need to do (1) to allow the testing to have a more permanent DB container for testing but I don't have the knowledge to say.

**Comment by knakatasf** [comment]
Hi @huss, thank you for your reply. Sorry, I mistyped. Currently, build.yml
- Runs a PostgreSQL database container. (1)
- Runs the tests locally, connecting with the database container. Testing is happening outside of Docker. (2)

travis.yml used to
- Install a PostgreSQL database locally. (3)
- Run the tests locally. connecting with the local database. Everything is happening outside of Docker. (4) 
- `docker compose up` for the web container and database container. (5)
- Run tests inside the containers. Everything is happening inside Docker. (6)

I believe build.yml wants to do at least (5)&(6) because this is how the developers builds the OED in their local machine. Do you want to keep (1)&(2) or (3)&(4) in build.yml?

**Comment by huss** [comment]
Hi @knakatasf Thanks for the update/reply. I'm open to any changes that make the process run better and/or faster as long it only impacts the GitHub Actions (not other installs) and gives the same result. So, do what you think and/or ask for any ideas.

**Comment by knakatasf** [comment]
Hi @huss, OK, I will keep the current workflow (1)&(2) and store its credentials (e.g., database password) in GitHub secrets. I will also create a separate yaml file for workflow (5)&(6), as this reflects how developers build the OED on their local machines.
Since these will be two separate workflow yaml files, they will run simultaneously and independently, meaning the execution time will not be longer than the current setup.

**Comment by Phinhas214** [comment]
Hi @huss, I would also like to work on this issue with @knakatasf.

**Comment by huss** [comment]
@Phinhas214 That is fine as long as @knakatasf is happy to work with you.

**Comment by Phinhas214** [comment]
> [@Phinhas214](https://github.com/Phinhas214) That is fine as long as [@knakatasf](https://github.com/knakatasf) is happy to work with you.

Yeah, I talked to him, and he said it was okay.

**Comment by knakatasf** [comment]
Hi @Phinhas214 
I've already finished (1)&(2). But why don't you update the build.yml file on your side too, and then we can compare our solutions.
For (5)&(6), I created a new workflow file: .github/workflows/docker-build.yml, along with docker-compose.ci.yml for this workflow. However, I'm currently facing an issue: when the web container tries to access the database container in the GitHub Actions VM, it crashes.
I think we can discuss this on Slack.

**Comment by Phinhas214** [comment]
Got it! I'll be able to start working on this on Friday (since I have finals for the next two days). I've also messaged you on slack.

**Comment by knakatasf** [comment]
Hi @huss, I’ve completed (1) & (2), but I’m stuck on (5) & (6). In the GitHub Actions VM, the database container starts and runs first, but when the web container tries to connect, authentication fails.

![Image](https://github.com/user-attachments/assets/9e148258-b8be-4256-8499-cd3a5877d24b)

I’ve confirmed that the web container is using the correct username and password:

<img width="902" alt="Image" src="https://github.com/user-attachments/assets/f2a59692-c9e7-459c-9ebd-2105799c2149" />

Could this issue be related to encryption? The database container requires SCRAM-SHA-256, but the web container might be using MD5 or plaintext authentication. However, these two containers run without issues on my local machine…

**Comment by huss** [comment]
@knakatasf Sorry for the delay in getting back to you. I'm not certain what is up. It seems it is using the expected connection for the DB. As you noted, it now uses a stronger encryption of scram-sha-256.

It is harder to say since the user and role and obscured. Also, when you verified the user/password were you running locally or checking the actual GitHub Action setup (I don't know how to do that)?

This isn't likely to be a lot of help. If you want me to look further at your precise work or do it together then let me know.

**Comment by knakatasf** [comment]
Hi @huss, sorry for the delay in replying. I noticed that I checked the user/password from the database container. I'm trying to reveal them from the web container, but they are always masked. It seems like the web container might be using the wrong user/password. I'll try to fix this. Thank you.
